src/simian/mac/api/groups.py

```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+#
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Groups API URL handlers."""
+
+import httplib
+import json
+import logging
+import webapp2
+
+from simian import settings
+from simian.mac import models
+from simian.mac.common import auth
+
+API_INFO_KEY = settings.API_INFO_KEY
+
+
+class GroupHandler(webapp2.RequestHandler):
+
+  def __init__(self, *args, **kwargs):
+    self.user = None
+    super(GroupHandler, self).__init__(*args, **kwargs)
+
+  def _oauth(self):
+    try:
+      self.user = auth.DoOAuthAuth()
+    except auth.NotAuthenticated:
+      enable_admin_check = True
+      # OAuth was either not used or failed, so perform regular user auth.
+      self.user = auth.DoUserAuth(is_admin=enable_admin_check)
+
+  def get(self):
+    """List groups, or list a group's memberhsip.
+
+    Send optional 'group' parameter to list membership of a specified group.
+    Otherwise a list of groups is returned.
+    """
+    key = self.request.get('key')
+
+    if not API_INFO_KEY:
+      logging.warning('API_INFO_KEY is unset; blocking all API info requests.')
+      self.response.set_status(httplib.UNAUTHORIZED)
+      return
+    elif key != API_INFO_KEY:
+      self.response.set_status(httplib.UNAUTHORIZED)
+      return
+
+    self.response.headers['Content-Type'] = 'application/json'
+
+    group = self.request.get('group')
+    if group:
+      existing_group = models.Group.get_by_key_name(group)
+      if existing_group:
+        self.response.out.write(json.dumps(existing_group.owners))
+    else:
+      self.response.out.write(json.dumps(models.Group.GetAllGroupNames()))
+
+  def post(self):
```

Comment:
This class is for the stateless API - which is the motivation behind this PR. I added the web interface bits exposing the groups feature as a courtesy to other Simian users.

I can't add a CSRF token because then this wouldn't be stateless. Instead I pass the existing API_INFO_KEY as a header (X-Simian-API-Info-Key) for all HTTP methods.